File: server/src/services/taskService.ts
import { prisma } from "../lib/prisma";
import { notify } from "./notifier";
import { emitToUsers } from "../realtime/io";

export interface AssignmentTarget {
  type: "USER" | "DEPARTMENT" | "ROLE_LEVEL";
  id?: string; // userId or departmentId
  level?: string; // for ROLE_LEVEL: MANAGER | DEPT_HEAD | ADMIN | MEMBER
  departmentId?: string; // optional scope for ROLE_LEVEL
}

export interface ResolvedAssignee {
  userId: string;
  via: "DIRECT" | "DEPARTMENT" | "ROLE_LEVEL";
  viaLabel: string | null;
}

/** Expand assignment targets (people, departments, role levels) to concrete users. */
export async function resolveAssignees(targets: AssignmentTarget[]): Promise<ResolvedAssignee[]> {
  const byUser = new Map<string, ResolvedAssignee>();

  for (const target of targets) {
    if (target.type === "USER" && target.id) {
      if (!byUser.has(target.id)) {
        byUser.set(target.id, { userId: target.id, via: "DIRECT", viaLabel: null });
      } else {
        // direct assignment wins over group-derived assignment
        byUser.set(target.id, { userId: target.id, via: "DIRECT", viaLabel: null });
      }
    } else if (target.type === "DEPARTMENT" && target.id) {
      const department = await prisma.department.findUnique({
        where: { id: target.id },
        include: { users: { where: { active: true }, select: { id: true } } },
      });
      if (!department) continue;
      for (const user of department.users) {
        if (!byUser.has(user.id)) {
          byUser.set(user.id, { userId: user.id, via: "DEPARTMENT", viaLabel: department.name });
        }
      }
    } else if (target.type === "ROLE_LEVEL" && target.level) {
      const users = await prisma.user.findMany({
        where: {
          active: true,
          roleLevel: target.level,
          ...(target.departmentId ? { departmentId: target.departmentId } : {}),
        },
        select: { id: true },
      });
      const label = target.level === "DEPT_HEAD" ? "Department heads" : `${target.level.charAt(0)}${target.level.slice(1).toLowerCase()}s`;
      for (const user of users) {
        if (!byUser.has(user.id)) {
          byUser.set(user.id, { userId: user.id, via: "ROLE_LEVEL", viaLabel: label });
        }
      }
    }
  }

  return [...byUser.values()];
}

export const taskInclude = {
  assigner: { select: { id: true, name: true, handle: true } },
  assignees: { include: { user: { select: { id: true, name: true, handle: true, departmentId: true } } } },
  sourceMessage: { select: { id: true, channelId: true, content: true } },
  meeting: { select: { id: true, title: true, startsAt: true } },
} as const;

export interface CreateTaskInput {
  title: string;
  description?: string;
  sourceMessageId?: string;
  dueDate?: Date | null;
  priority?: string;
  meetingId?: string;
  targets: AssignmentTarget[];
}

export async function createTask(assignerId: string, input: CreateTaskInput) {
  const assignees = await resolveAssignees(input.targets);
  if (assignees.length === 0) throw new Error("Task must have at least one assignee");

  const task = await prisma.task.create({
    data: {
      title: input.title,
      description: input.description,
      sourceMessageId: input.sourceMessageId,
      dueDate: input.dueDate ?? null,
      priority: input.priority ?? "MEDIUM",
      meetingId: input.meetingId,
      assignerId,
      assignees: {
        create: assignees.map((a) => ({ userId: a.userId, via: a.via, viaLabel: a.viaLabel })),
      },
    },
    include: taskInclude,
  });

  const assigner = task.assigner;
  const due = task.dueDate ? ` (due ${task.dueDate.toISOString().slice(0, 10)})` : "";
  await notify(
    assignees.map((a) => a.userId).filter((id) => id !== assignerId),
    {
      type: "task_assigned",
      title: `${assigner.name} assigned you a task: ${task.title}`,
      body: `Priority ${task.priority}${due}`,
      link: `/tasks?taskId=${task.id}`,
    }
  );

  emitTaskUpdate(task);
  return task;
}

export function emitTaskUpdate(task: { id: string; assignerId: string; assignees: { userId: string }[] }) {
  const userIds = [task.assignerId, ...task.assignees.map((a) => a.userId)];
  emitToUsers(userIds, "task:updated", { taskId: task.id });
}

File: dotnet/TeamCollabWebForms/App_Code/TaskService.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;

public static class TaskService
{
    public static readonly string[] Statuses = { "NOT_STARTED", "DECISION_MAKING", "IN_PROGRESS", "COMPLETED" };

    public static string StatusLabel(string status)
    {
        switch (status)
        {
            case "NOT_STARTED": return "Not yet started";
            case "DECISION_MAKING": return "Decision making";
            case "IN_PROGRESS": return "In progress";
            case "COMPLETED": return "Completed";
            default: return status;
        }
    }

    /// <summary>Expand person/department/role-level targets into concrete assignees.</summary>
    public static Dictionary<int, string[]> ResolveAssignees(
        IEnumerable<int> userIds, IEnumerable<int> departmentIds, IEnumerable<string> roleLevels)
    {
        // userId -> { Via, ViaLabel }
        var assignees = new Dictionary<int, string[]>();

        foreach (var deptId in departmentIds ?? Enumerable.Empty<int>())
        {
            var deptName = Db.Scalar("SELECT Name FROM Departments WHERE Id = @d", Db.P("@d", deptId)) as string;
            if (deptName == null) continue;
            foreach (DataRow r in Db.Query(
                "SELECT Id FROM Users WHERE Active = 1 AND DepartmentId = @d", Db.P("@d", deptId)).Rows)
            {
                var id = (int)r["Id"];
                if (!assignees.ContainsKey(id)) assignees[id] = new[] { "DEPARTMENT", deptName };
            }
        }

        foreach (var level in roleLevels ?? Enumerable.Empty<string>())
        {
            var label = level == "DEPT_HEAD" ? "Department heads"
                      : char.ToUpper(level[0]) + level.Substring(1).ToLowerInvariant() + "s";
            foreach (DataRow r in Db.Query(
                "SELECT Id FROM Users WHERE Active = 1 AND RoleLevel = @r", Db.P("@r", level)).Rows)
            {
                var id = (int)r["Id"];
                if (!assignees.ContainsKey(id)) assignees[id] = new[] { "ROLE_LEVEL", label };
            }
        }

        foreach (var userId in userIds ?? Enumerable.Empty<int>())
        {
            assignees[userId] = new[] { "DIRECT", null }; // direct assignment wins
        }

        return assignees;
    }

    public static int Create(int assignerId, string assignerName, string title, string description,
        DateTime? dueDate, string priority, int? sourceMessageId, int? meetingId,
        Dictionary<int, string[]> assignees)
    {
        if (assignees.Count == 0) throw new InvalidOperationException("A task needs at least one assignee.");

        var taskId = Db.Insert(
            @"INSERT INTO Tasks (Title, Description, SourceMessageId, AssignerId, Priority, DueDate, MeetingId)
              VALUES (@title, @desc, @src, @assigner, @priority, @due, @meeting)",
            Db.P("@title", title), Db.P("@desc", description), Db.P("@src", sourceMessageId),
            Db.P("@assigner", assignerId), Db.P("@priority", priority ?? "MEDIUM"),
            Db.P("@due", dueDate), Db.P("@meeting", meetingId));

        foreach (var pair in assignees)
        {
            Db.Exec(
                @"INSERT INTO TaskAssignees (TaskId, UserId, Via, ViaLabel) VALUES (@t, @u, @via, @label)",
                Db.P("@t", taskId), Db.P("@u", pair.Key),
                Db.P("@via", pair.Value[0]), Db.P("@label", pair.Value[1]));
        }

        var due = dueDate.HasValue ? " (due " + dueDate.Value.ToString("yyyy-MM-dd HH:mm") + ")" : "";
        NotificationService.Notify(
            assignees.Keys.Where(id => id != assignerId),
            "task_assigned",
            assignerName + " assigned you a task: " + title,
            "Priority " + (priority ?? "MEDIUM") + due,
            "Tasks.aspx");

        return taskId;
    }

    public static void UpdateStatus(int taskId, string status, int actorId, string actorName)
    {
        if (!Statuses.Contains(status)) throw new InvalidOperationException("Invalid status: " + status);

        Db.Exec(
            @"UPDATE Tasks SET Status = @s,
                     CompletedAt = CASE WHEN @s = 'COMPLETED' THEN COALESCE(CompletedAt, SYSUTCDATETIME()) ELSE NULL END,
                     UpdatedAt = SYSUTCDATETIME()
              WHERE Id = @id", Db.P("@s", status), Db.P("@id", taskId));

        var task = Db.Query("SELECT Title, AssignerId FROM Tasks WHERE Id = @id", Db.P("@id", taskId)).Rows[0];
        var interested = new List<int> { (int)task["AssignerId"] };
        foreach (DataRow r in Db.Query("SELECT UserId FROM TaskAssignees WHERE TaskId = @id", Db.P("@id", taskId)).Rows)
            interested.Add((int)r["UserId"]);

        NotificationService.Notify(
            interested.Where(id => id != actorId),
            "task_updated",
            actorName + " moved \"" + task["Title"] + "\" to " + StatusLabel(status).ToLowerInvariant(),
            null, "Tasks.aspx");
    }

    public static void Acknowledge(int taskId, int userId)
    {
        Db.Exec(
            @"UPDATE Tasks SET AcknowledgedAt = COALESCE(AcknowledgedAt, SYSUTCDATETIME())
              WHERE Id = @id AND EXISTS (SELECT 1 FROM TaskAssignees WHERE TaskId = @id AND UserId = @u)",
            Db.P("@id", taskId), Db.P("@u", userId));
    }
}
